pkg/registry/populator_test.go

```diff
@@ -1534,3 +1534,425 @@ func TestSemverPackageManifest(t *testing.T) {
 		})
 	}
 }
+
+func TestSubstitutesFor(t *testing.T) {
+	type args struct {
+		bundles  []string
+		rebuilds []string
+	}
+	type expected struct {
+		bundles        []string
+		channels       []string
+		channelHeads   map[string]string
+		defaultChannel string
+	}
+	tests := []struct {
+		description string
+		args        args
+		expected    expected
+	}{
+		{
+			description: "SubstitutesFor/FirstSubstitutionReplaces",
+			args: args{
+				bundles: []string{
+					"prometheus.0.22.2",
+					"prometheus.0.14.0",
+					"prometheus.0.15.0",
+				},
+				rebuilds: []string{
+					"prometheus.0.15.0.substitutesfor",
+				},
+			},
+			expected: expected{
+				bundles: []string{
+					"quay.io/test/prometheus.0.22.2",
+					"quay.io/test/prometheus.0.14.0",
+					"quay.io/test/prometheus.0.15.0.substitutesfor",
+					"quay.io/test/prometheus.0.15.0.substitutesfor",
+					"quay.io/test/prometheus.0.15.0",
+					"quay.io/test/prometheus.0.15.0",
+					"quay.io/test/prometheus.0.14.0",
+				},
+				channels: []string{
+					"stable",
+					"preview",
+					"preview",
+					"stable",
+					"stable",
+					"preview",
+					"preview",
+				},
+				channelHeads: map[string]string{
+					"preview": "prometheusoperator.0.22.2",
+					"stable":  "prometheusoperator.0.15.0+1-freshmaker-rebuild",
+				},
+				defaultChannel: "preview",
+			},
+		},
+		{
+			description: "SubstitutesFor/FirstSubstitutionReplacesNonHead",
+			args: args{
+				bundles: []string{
+					"prometheus.0.22.2",
+					"prometheus.0.14.0",
+					"prometheus.0.15.0",
+				},
+				rebuilds: []string{
+					"prometheus.0.14.0.substitutesfor",
+				},
+			},
+			expected: expected{
+				bundles: []string{
+					"quay.io/test/prometheus.0.22.2",
```

Comment:
generating these fixtures at runtime could be preferable. not sure where others stand on this.

Reply:
what do you mean?

Reply:
I think nick's point is that the "substitutes-for" variants are 1 line different from the other manifests, we could create these fixtures by loading one of the existing fixtures and mutating it.

Reply:
Yes, in that vein. I see two appealing benefits:

1. Reduce the footprint of our test fixtures
2. Make it clear to the test reader how the fixture has been perturbed